File: algorithms/solutions/week18/toanha/min_stack.ex
defmodule MinStack do
  @max_int 2_147_483_647
  defstruct arr: [], min: @max_int

  def push(model = %__MODULE__{min: min}, value) when value > min do
    %{model | arr: [value | model.arr]}
  end

  def push(model = %__MODULE__{min: min}, value) when value <= min do
    %{model | arr: [value | model.arr], min: value}
  end

  def pop(model = %__MODULE__{arr: arr = [h | t]}) when length(arr) >= 2 do
    min = Enum.min(t)
    %{model | arr: t, min: min}
  end

  def pop(model = %__MODULE__{arr: arr}) when length(arr) < 2 do
    %MinStack{}
  end

  def top(model = %__MODULE__{arr: arr}) when length(arr) == 0 do
    0
  end

  def top(model = %__MODULE__{arr: arr = [h | t]}) when length(arr) > 0 do
    h
  end

  def get_min(model = %__MODULE__{arr: arr}) when length(arr) == 0 do
    0
  end

  def get_min(model = %__MODULE__{arr: arr, min: min}) when length(arr) > 0 do
    min
  end
end

defmodule Main do
  def run() do
    stack = %MinStack{}

    stack
    |> MinStack.push(-2)
    |> IO.inspect()
    |> MinStack.push(0)
    |> IO.inspect()
    |> MinStack.push(-3)
    |> IO.inspect()
    |> MinStack.pop()
    |> IO.inspect()
    |> MinStack.top()
    |> IO.inspect()

    # |> MinStack.get_min()
    # |> IO.inspect()
  end
end

Main.run()

File: algorithms/solutions/week18/qcam.solutions/min_stack.rb
class Stack
  def initialize()
    @items = []
    @size = 0
  end

  def push(item)
    @items[size()] = item
    @size = size() + 1
  end

  def pop()
    if empty?()
      raise "failed to pop stack, stack is empty"
    else
      item = items()[size() - 1]
      @items.delete_at(size() - 1)
      @size = size() - 1

      return item
    end
  end

  def top()
    return nil if empty?()

    items()[size() - 1]
  end

  private

  attr_reader :size, :items

  def empty?()
    size() == 0
  end
end

class MinStack
  def initialize()
    @actual_stack = Stack.new()
    @min_stack = Stack.new()
  end

  def push(item)
    @actual_stack.push(item)

    min = min_stack().top()

    if min.nil?() || min >= item
      @min_stack.push(item)
    end
  end

  def pop()
    item = @actual_stack.pop()

    if item == min_stack().top()
      @min_stack.pop()
    end

    return item
  end

  def top()
    actual_stack().top()
  end

  def get_min()
    min_stack().top()
  end

  private

  attr_reader :min_stack, :actual_stack
end
